[MLIR][NVVM] Add globaltimer_lo support in NVVM Dialect and NVPTX backend

DharuniRAcharya opened the pull request:
This patch adds support for reading the global timer low register in the NVVM dialect and NVPTX backend. 
This change includes addition of `NVVM_GlobalTimerLoOp` operation to NVVM dialect and `int_nvvm_read_ptx_sreg_globaltimer_lo` intrinsic to NVPTX backend.
All the lit tests have been added.

DharuniRAcharya committed:
[MLIR][NVPTX] Expose globaltimer_lo in NVVM Dialect and Add corresponding intrinsics in NVPTX backend

durga4github commented:
Dharuni, Kindly resolve the conflicts and push once, so that we can get build results.
Also, please add the link to the PTX instruction in the commit message.

DharuniRAcharya committed:
[MLIR][NVVM] Add globaltimer_lo support in NVVM Dialect and NVPTX backend
This patch adds support for reading the global timer low register in the NVVM dialect and NVPTX backend.
This change includes addition of NVVM_GlobalTimerLoOp operation to NVVM dialect and int_nvvm_read_ptx_sreg_globaltimer_lo intrinsic to NVPTX backend.
All the lit tests have been added.
Adding the relevant PTX link here for reference:
https://docs.nvidia.com/cuda/parallel-thread-execution/#special-registers-globaltimer

DharuniRAcharya committed:
Merge branch 'main' into dev/dharunira/upstream-globaltimer-lo

DharuniRAcharya committed:
Merge branch 'main' into dev/dharunira/upstream-globaltimer-lo

gonzalobg reviewed (approved):
I think this is fine as long as eventually: 

```
%0 = call i64 nvvm.read.ptx.sreg.globaltimer()
%1 = trunc i64 %0 to i32
```
eventually lowers to `%globaltimer_lo` in PTX (e.g. by being transformed to `nvvm.read.ptx.sreg.globaltimer_lo()`), but that can be done later.

grypp commented on llvm/test/CodeGen/NVPTX/intrinsics.ll:
```diff
@@ -267,6 +267,23 @@ define i64 @test_globaltimer() {
+; CHECK-NEXT:    add.s32 %r3, %r1, %r2;
+; CHECK-NEXT:    st.param.b32 [func_retval0], %r3;
+; CHECK-NEXT:    ret;
+  %a = tail call i32 @llvm.nvvm.read.ptx.sreg.globaltimer.lo()
+  %b = tail call i32 @llvm.nvvm.read.ptx.sreg.globaltimer.lo()
+  %ret = add i32 %a, %b
```
It's better to split llvm and mlir PRs. It's okay for now
